docs/RFCS/distributed_sql.md

```diff
@@ -0,0 +1,1472 @@
+follows from that.
+
+One place where physical planning involves difficult decisions is when
+distributing the second stage of an aggregation or join - we could set up any
+number of "buckets" (and subsequent flows) on any nodes. E.g. see the `summer`
+example. Fortunately there is a simple strategy we can start with - use as many
+buckets as input flows and distribute them among the same nodes. This strategy
+scales well with the query size: if a query draws data from a single node, we
+will do all the aggregation on that node; if a query draws data from many nodes,
+we will distribute the aggregation among those nodes. 
+
+A "stage 2" refinement to this strategy is detecting when a computation (and
+subsequent stages) might be cheap enough to not need distribution, in which case
+we can perform the aggregation on the gateway node. Further improvements
+(statistics based) can be investigated later.
+
+## Processor infrastructure and implementing processors
+
+This involves building the framework within which we can run processors and
+implementing the various processors we need. We can start with the table readers
+(enough for M1). If necessary, this work stream can advance faster than the
+logical planning stream - we can build processors even if the logical plan isn't
+using them yet (as long as there is a good testing framework in place); we can
+also potentially use the implementations internally, hidden behind a `planNode`
+interface and running non-distributed.
+
+### Joins
+
+A tangential work stream is to make progress toward supporting joins (initially
+non-distributed). This involves building the processor that will be at the heart
+of the hash join implementation and integrating that code with the current
+`planNode` tree.
+
+## Scheduling
+
+The problem of efficiently queuing and scheduling processors will also involve a
+long tail of improvements. But we can start with a basic implementation using
+simple strategies:
+ - the queue ordering is by transactions; we don't order individual processors
+ - limit the number of transactions for which we run processors at any one time;
+   we can also limit the total number of processors running at any one time, as
+   long as we allow all the processors needed for at least one txn
+ - the txn queuing order is a function of the txn timestamp and its priority,
+   allowing the nodes to automatically agree on the relative ordering of
+   transactions, eliminating deadlock scenarios (example of a deadlock: txn A
+   has some processors running on node 1, and some processors on node 2 that are
+   queued behind running processors of txn B; and txn B also has some processors
+   that are queued behind txn A on node 1)
+
+We won't need anything fancier in this area to reach M1 and M2.
+
+## KV integration
+
+KV integration involves three aspects:
+
+1. Range information lookup
+
+   At the physical planning stage we need to break up key spans into ranges and
```

Comment:
We might also use the range info for a table during the logical planning stage for index selection and join ordering. In addition to the number of ranges which provides an estimate of the size of a table/index, the range start and end keys provide a regular sampling of the associated table/index data.